### ISSUE TITLE
Idea: Muti-fix by box selection in Repair mode

### ISSUE THREAD

**Issue by dnqbob** [OpenRA/OpenRA#17339]
I make a GIF about this function in my expectation:
![demo opra repair mode](https://user-images.githubusercontent.com/13763394/68990839-da5d0c80-0892-11ea-8b14-3a12326dd8a3.gif)

In some mission, players just get a bunch of heavily damaged buildings (so called "abandoned base", etc.) requires repairing. Clicking one by one can be annoying.

When players build up large base defenses and heavy vehicles as meat shield to encounter a large wave of enemies, it is hard to distinguish which one is damaged during the battle, and after the battle. It is even harder to get each one of them repaired by using simple keyboard/mouse input.

After a strike from support power or super weapon in the bases, we need a more convenient way to fix damaged buildings which is as quickly as AI players do, and then focus attention on battle ahead instead of which building need to be clicked on.

I think this idea, box selection in repair mode, can repair these flaws that original C&Cs have. 

Probably we can also online/offline buildings or defenses like this. But emmmm, not for sale.

**Comment by dragunoff** [comment]
This would also make aircraft easier to repair because the player can just drag a box around a moving aircraft instead of trying to click precisely on it.

This however raises the question of whether other modes (power down, sell) should work with select-dragging.

**Comment by dnqbob** [comment]
> This would also make aircraft easier to repair because the player can just drag a box around a moving aircraft instead of trying to click precisely on it.

Yep, especially for fixed-wing aircrafts that enjoy circling in the sky.

> This however raises the question of whether other modes (power down, sell) should work with select-dragging.

As for power mode I believe select-dragging can be good, but there are some situations needing taking care of by devs, such as what should the game-platform do when selecting box contains both offline and online buildings.

As for selling mode, I would prefer not, becuase selling in Openra cannot be undone and very dangrous with a mistaken select-dragging.

**Comment by GraionDilach** [comment]
I would prefer if all these would be controllable by modders if they want these features or not. Some modders might have unit selling where group sell is worthy (for example YR and group selling would send them all into a Grinder).

**Comment by dnqbob** [comment]
> I would prefer if all these would be controllable by modders if they want these features or not. Some modders might have unit selling where group sell is worthy (for example YR and group selling would send them all into a Grinder).

It is always good when modders can control the feature as detailedly as possible without recompile the entire platform. However, for something about input&controll we still have to look into the source code for now.

**Comment by dnqbob** [comment]
OK dudes, I have already made this function in Release-20191117. 

GIF related.
![in game performance(real)](https://user-images.githubusercontent.com/13763394/69944566-9caee380-1522-11ea-8e5b-be75cce2eb44.gif)

Still a little flaws remain. 

I don't want to give "OrderGenerater" some of "Wiget" function, and it is beyond my ability to give "OrderGenerater" a  dynamic mouse-following function without break encapsulation, or change too much of orignal design. So you can see in GIF, when dragging action happens, the cursor returns to normal cursor(only in visual), but still in Repair Mode, for the dragging rectangle is in fact drawing by "WorldInteractionControllerWidget" dynamicly in visual. I don't know if you guys can accept this :(

**Comment by GraionDilach** [comment]
This sounds horrible even without seeing the code.

You should base this feature on #16999 in bleed.

**Comment by tovl** [comment]
You can also take a look at `OpenRa.Mods.Common.Traits.SelectDirectionalTarget` or #17099 to see how similar features are implemented.

**Comment by dnqbob** [comment]
> This sounds horrible even without seeing the code.
> You should base this feature on #16999 in bleed.

> You can also take a look at OpenRa.Mods.Common.Traits.SelectDirectionalTarget or #17099 to see how similar features are implemented.

Thanks, devs. I will take a look at that but I am not sure if I will continue on this (I am a newbie for C# and OpenRA).

**Comment by dnqbob** [comment]
tovl, I fail to complie your mouse-dragging function branch all the time so I guess I can not use it as base branch, but thank you anyway.

In addtion, I have already found the reason why the cursor image will become normal cursor arrow when the white rectangle is drawing.

It is due to the "GetCursor()" in "WorldInteractionControllerWidget" should work when "Draw()" in this widget like the others, but it is apparently not. I can't find which class in codes control the "Draw()", so I guess this is it.

But it effects little to actual gameplay. I decided to commit my branch to bleed.

**Comment by dnqbob** [comment]
Deeply grateful to @tovl 's help, now the bug is solved and modified repair mode on my branch is working smoothly at present. 

![Cursor Dragging](https://user-images.githubusercontent.com/13763394/70299568-32eb4e00-1830-11ea-96b1-a799b226c861.gif)

Should I close the issue or talk about other function like online/offline mode here?
